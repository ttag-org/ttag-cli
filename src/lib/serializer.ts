import { po } from "gettext-parser";
import { PoData } from "../lib/parser";

export type SerializeOptions = {
    foldLength?: number;
};

export function serialize(poData: PoData, options?: SerializeOptions): Buffer {
    return po.compile(poData, options);
}
