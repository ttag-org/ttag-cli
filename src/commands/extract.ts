import * as ora from "ora";
import * as fs from "fs";
import * as c3poTypes from "../types";
import { extractAll } from "../lib/extract";

async function extract(
    output: string,
    paths: string[],
    lang: string = "en",
    ttagOverrideOpts?: c3poTypes.TtagOpts,
    ttagRcOpts?: c3poTypes.TtagRc,
    exclude?: string[]
) {
    const progress: c3poTypes.Progress = ora(
        `[ttag] extracting translations to ${output} ...`
    );
    progress.start();

    const excludeRegexp = exclude ? new RegExp(exclude.join("|")) : undefined;

    const result = await extractAll(
        paths,
        lang,
        progress,
        ttagOverrideOpts,
        ttagRcOpts,
        excludeRegexp
    );
    fs.writeFileSync(output, result);
    progress.succeed(`[ttag] translations extracted to ${output}`);
}

export default extract;
