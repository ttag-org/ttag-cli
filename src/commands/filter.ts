import { parse, PoData, Message, Messages, Translations } from "../lib/parser";
import { serialize } from "../lib/serializer";
import * as fs from "fs";

enum RuleType {
    Must,
    MustNot
}

interface PoFilterParams {
    fuzzy?: boolean;
    translated?: boolean;
    referenceRe?: RegExp;
}

type TestFunction = (msg: Message) => boolean;

function FuzzyTest(msg: Message): boolean {
    if (msg.comments == undefined) {
        return false;
    }
    return msg.comments.flag && msg.comments.flag.includes("fuzzy");
}

function TranslatedTest(msg: Message): boolean {
    return msg.msgstr.filter(s => s.length > 0).length == msg.msgstr.length;
}

function ReferenceReTest(msg: Message, re: RegExp): boolean {
    return (
        msg.comments != undefined &&
        msg.comments.reference != undefined &&
        re.test(msg.comments.reference)
    );
}

type FilterRules = Array<[TestFunction, RuleType]>;

class PoFilter {
    fuzzy?: boolean;
    translated?: boolean;
    referenceRe?: RegExp;

    constructor({ fuzzy, translated, referenceRe }: PoFilterParams) {
        this.fuzzy = fuzzy;
        this.translated = translated;
        this.referenceRe = referenceRe;
    }

    /* set fuzzy flag */
    withFuzzy() {
        return new PoFilter(Object.assign({}, this, { fuzzy: true }));
    }

    /* set no fuzzy flag */
    withoutFuzzy() {
        return new PoFilter(Object.assign({}, this, { fuzzy: false }));
    }

    /* set translation flag */
    withTranslation() {
        return new PoFilter(Object.assign({}, this, { translated: true }));
    }

    /* set no translation flag */
    withoutTranslation() {
        return new PoFilter(Object.assign({}, this, { translated: false }));
    }

    withReferenceRe(referenceRe: RegExp) {
        return new PoFilter(
            Object.assign({}, this, { referenceRe: referenceRe })
        );
    }

    /* build rule chain according to state and apply to translations */
    apply(translations: Translations): Translations {
        const rules = <FilterRules>[];
        if (this.fuzzy == true) {
            rules.push([FuzzyTest, RuleType.Must]);
        }
        if (this.fuzzy == false) {
            rules.push([FuzzyTest, RuleType.MustNot]);
        }
        if (this.translated == true) {
            rules.push([TranslatedTest, RuleType.Must]);
        }
        if (this.translated == false) {
            rules.push([TranslatedTest, RuleType.MustNot]);
        }
        const reg = this.referenceRe;
        if (reg !== undefined) {
            rules.push([
                (m: Message) => ReferenceReTest(m, reg),
                RuleType.Must
            ]);
        }
        const newTranslations = <Translations>{};
        for (let [ctxt, messages] of filterTranslationsStream(
            translations,
            rules
        )) {
            newTranslations[ctxt] = messages;
        }
        return newTranslations;
    }
}

/* Test rule according to type */
function testRule(test: TestFunction, rule: RuleType, msg: Message): boolean {
    switch (rule) {
        case RuleType.Must: {
            return test(msg);
        }
        case RuleType.MustNot: {
            return !test(msg);
        }
    }
}

/* Test each message with tester according to rule */
function* filterMessagesStream(
    messages: Messages,
    rules: FilterRules
): IterableIterator<Message> {
    for (const msgid of Object.keys(messages)) {
        const msg = messages[msgid];
        if (msgid == "") {
            // skip empty message id
            yield msg;
        }

        let allPassed = true;
        for (let [test, rule] of rules) {
            if (!testRule(test, rule, msg)) {
                allPassed = false;
                break;
            }
        }
        if (allPassed) {
            yield msg;
        }
    }
}

/* Run all messages by context through filter stream */
function* filterTranslationsStream(
    translations: Translations,
    rules: FilterRules
): IterableIterator<[string, Messages]> {
    for (const contextKey of Object.keys(translations)) {
        const context = translations[contextKey];
        const newContext = <Messages>{};
        for (const msg of filterMessagesStream(context, rules)) {
            newContext[msg.msgid] = msg;
        }
        if (Object.keys(newContext).length > 0) {
            yield [contextKey, newContext];
        }
    }
}

export default function filter(
    path: string,
    fuzzy: boolean,
    noFuzzy: boolean,
    translated: boolean,
    notTranslated: boolean,
    referenceRe: string
) {
    if (fuzzy && noFuzzy) {
        throw "Choose one of fuzzy or no-fuzzy args";
    }
    if (translated && notTranslated) {
        throw "Choose one of translated or not translated args";
    }
    if (referenceRe) {
        try {
            new RegExp(referenceRe);
        } catch {
            throw "Invalid regular expression for reference";
        }
    }
    let filter = new PoFilter({});
    if (fuzzy) {
        filter = filter.withFuzzy();
    }
    if (noFuzzy) {
        filter = filter.withoutFuzzy();
    }
    if (translated) {
        filter = filter.withTranslation();
    }
    if (notTranslated) {
        filter = filter.withoutTranslation();
    }
    if (referenceRe) {
        filter = filter.withReferenceRe(new RegExp(referenceRe));
    }

    const poData = parse(fs.readFileSync(path).toString());

    const filteredPoData = <PoData>{
        headers: poData.headers,
        translations: filter.apply(poData.translations)
    };
    process.stdout.write(serialize(filteredPoData));
}
