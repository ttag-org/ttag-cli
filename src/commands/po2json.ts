import * as ora from "ora";
import * as ttagTypes from "../types";
import { parse, PoDataCompact, PoData } from "../lib/parser";
import { iterateTranslations, convert2Compact } from "../lib/utils";
import { checkDuplicateKeys } from "../lib/checkDuplicateKeys";
import * as fs from "fs";

export default function po2json(
    path: string,
    pretty: boolean,
    nostrip: boolean,
    format: "compact" | "verbose"
) {
    const progress: ttagTypes.Progress = ora(
        `[ttag] po2json translation from ${path} ...`
    );
    let poData: PoData | PoDataCompact = parse(
        fs.readFileSync(path).toString()
    );
    const errMessage = checkDuplicateKeys(poData);

    if (errMessage) {
        progress.fail(errMessage);
        process.exit(1);
    }
    const messages = iterateTranslations(poData.translations);
    if (!nostrip) {
        const header = messages.next().value;
        delete header.comments;
        for (const msg of messages) {
            delete msg.comments;
        }
    }
    if (format === "compact") {
        poData = convert2Compact(poData);
    }
    process.stdout.write(JSON.stringify(poData, null, pretty ? 2 : 0));
}
