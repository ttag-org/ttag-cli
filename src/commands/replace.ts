import "../declarations";
import * as ora from "ora";
import * as c3poTypes from "../types";
import { makeBabelConf } from "../defaults";
import * as babel from "@babel/core";
import * as path from "path";
import * as fs from "fs";
import { TransformFn, pathsWalk } from "../lib/pathsWalk";
import * as mkdirp from "mkdirp";

async function replace(
    pofile: string,
    out: string,
    srcPath: string,
    overrideOpts?: c3poTypes.TtagOpts
) {
    const progress: c3poTypes.Progress = ora(
        `[ttag] replacing source files with translations ...`
    );
    progress.start();
    let ttagOpts: c3poTypes.TtagOpts = {
        resolve: { translations: pofile }
    };

    if (overrideOpts) {
        ttagOpts = Object.assign(ttagOpts, overrideOpts);
    }
    const babelOptions = makeBabelConf(ttagOpts);
    const transformFn: TransformFn = async file => {
        const relativePath = path.relative(srcPath, file);
        const resultPath = path.join(out, relativePath);
        const result = await babel.transformFileAsync(file, babelOptions);
        const dir = path.dirname(resultPath);
        if (dir !== ".") {
            mkdirp.sync(dir);
        }
        if (!result) {
            progress.fail("Failed to replace");
            return;
        }
        await fs.promises.writeFile(resultPath, result.code);
    };

    await pathsWalk([srcPath], progress, transformFn);
    progress.succeed(`[ttag] replace is done`);
}

export default replace;
