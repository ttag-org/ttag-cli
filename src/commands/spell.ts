import * as fs from "fs";
import { parse } from "../lib/parser";
import * as ora from "ora";
import * as c3poTypes from "../types";
import chalk from "chalk";
import { iterateTranslations } from "../lib/utils";
import { getChecker, Checker } from "../lib/spell";
import { printMsg } from "../lib/print";

const cleanRe = new RegExp(/[;:,.?"'!\(\)«»]/g);

export default function spell(path: string, locale?: string) {
    // Force color output even on tty, otherwise this command is useless
    chalk.enabled = true;
    chalk.level = 1;

    const data = fs.readFileSync(path).toString();
    const poData = parse(data);
    locale = locale || poData.headers.language || poData.headers.Language;
    if (!locale) {
        console.log("Cannot detect locale from pofile, please provide it");
        return;
    }
    const loadProgress: c3poTypes.Progress = ora(
        `Loading dict for ${locale}...`
    );
    loadProgress.start();
    getChecker(locale).then(function(checker: Checker) {
        loadProgress.succeed(`${locale} dict loaded`);
        const checkProgress: c3poTypes.Progress = ora(
            `Checking pofile ${path}...\n\n`
        );
        checkProgress.start();
        const messages = iterateTranslations(poData.translations);
        messages.next(); // skip headers
        for (const msg of messages) {
            let hasErrors = false;
            for (let i = 0; i < msg.msgstr.length; i++) {
                for (const word of msg.msgstr[i].split(" ")) {
                    const cleanWord = word.replace(cleanRe, "");
                    if (cleanWord && !checker.check(cleanWord)) {
                        msg.msgstr[i] = msg.msgstr[i].replace(
                            cleanWord,
                            chalk.underline.bgRed(cleanWord)
                        );
                        hasErrors = true;
                    }
                }
            }
            if (hasErrors) {
                printMsg(msg);
                console.log("\n");
            }
        }
        checkProgress.succeed(`${path} checked`);
    });
}
