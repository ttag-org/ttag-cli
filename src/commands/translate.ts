import * as fs from "fs";
import * as readlineSync from "readline-sync";
import chalk from "chalk";
import { parse, Translations, PoData } from "../lib/parser";
import { serialize } from "../lib/serializer";
import {
    printComments,
    printContext,
    printMsgid,
    printMsgidPlural
} from "../lib/print";

/* Generate untranslated messages along with context */
export function* untranslatedStream(translations: Translations): any {
    for (const contextKey of Object.keys(translations)) {
        const context = translations[contextKey];
        for (const msgid of Object.keys(context)) {
            const msg = context[msgid];
            if (msg.msgstr.includes("") || !msgid) {
                msg.msgstr = yield [contextKey, msg];
            }
        }
    }
}

/* Read file and parse it shorthand */
export function read(path: string): PoData {
    return parse(fs.readFileSync(path).toString());
}

/* Stream translations for each form if many(plural) */
function* translationStream(msgstr: string[]): IterableIterator<string> {
    if (msgstr.length > 1) {
        for (let i = 0; i < msgstr.length; i++) {
            yield readlineSync.question(chalk.yellow(`msgstr[${i}]: `));
        }
    } else {
        yield readlineSync.question(chalk.yellow(`msgstr: `));
    }
}

export default function translate(path: string, output: string) {
    const poData = read(path);
    const stream = untranslatedStream(poData.translations);
    // skip first message(empty msgid in header)
    stream.next();
    var { value, done } = stream.next("");
    while (!done) {
        let [ctxt, msg] = value;
        printComments(msg.comments);
        printContext(ctxt);
        printMsgid(msg.msgid);
        printMsgidPlural(msg.msgid_plural);
        const translation = Array.from(translationStream(msg.msgstr));
        const data = stream.next(translation);
        [value, done] = [data.value, data.done];
        console.log();
    }
    fs.writeFileSync(output, serialize(poData));
    console.log(`Translations written to ${output}`);
}
