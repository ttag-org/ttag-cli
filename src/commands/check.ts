/*
Compare designated pofile with pot file extracted from all files in all paths
*/

import * as ora from "ora";
import * as fs from "fs";
import { extractAll } from "../lib/extract";
import { checkDuplicateKeys } from "../lib/checkDuplicateKeys";
import * as c3poTypes from "../types";
import { parse, PoData } from "../lib/parser";

/*
Run any string in stream through warning first
*/
function* warningPipe(
    pofile: string,
    progress: c3poTypes.Progress,
    stream: IterableIterator<string>
): IterableIterator<string> {
    for (const str of stream) {
        progress.warn(`Translation '${str}' is not found in ${pofile}`);
        yield str;
    }
}

/*
Unpack pofile into contextKey/messageId pairs
*/
function* unpackPoData(poData: PoData): IterableIterator<[string, string]> {
    for (const contextKey of Object.keys(poData.translations)) {
        for (const msgid of Object.keys(poData.translations[contextKey])) {
            const keyMsg = poData.translations[contextKey][msgid];
            yield [contextKey, keyMsg.msgid];
        }
    }
}

/*
Find untranslated string by extracting from translations(pofile)
using key/context from keys(pot file).
*/
function* getUntranslated(
    translations: PoData,
    keysOnly: PoData
): IterableIterator<string> {
    for (const [contextKey, msgid] of unpackPoData(keysOnly)) {
        const context = translations.translations[contextKey];
        if (!context || !context[msgid]) {
            yield msgid;
            continue;
        }
        const msgstr = context[msgid].msgstr;
        if (msgstr.filter(s => !!s).length == 0) {
            yield msgid;
        }
    }
}

/*
Check all keys from pots(keys only files) are present in pofile(files with translations)
*/
async function check(
    pofile: string,
    paths: string[],
    lang: string,
    overrideOpts?: c3poTypes.TtagOpts,
    ttagRcOpts?: c3poTypes.TtagRc,
    skip?: "translation"
) {
    const progress: c3poTypes.Progress = ora(
        `[ttag] checking translations from ${paths} ...`
    );
    // progress.start();

    const translations = parse(fs.readFileSync(pofile).toString());
    const keysOnly = parse(
        await extractAll(paths, lang, progress, overrideOpts, ttagRcOpts)
    );

    let untranslatedStream = getUntranslated(translations, keysOnly);
    untranslatedStream = warningPipe(pofile, progress, untranslatedStream);
    const untranslated = Array.from(untranslatedStream);

    const errMessage = checkDuplicateKeys(translations);

    if (errMessage) {
        progress.fail(errMessage);
        process.exit(1);
    }

    if (untranslated.length && skip !== "translation") {
        progress.fail(
            `[ttag] has found ${untranslated.length} untranslated string(s)`
        );
        process.exit(1);
    } else {
        // progress.succeed(`[ttag] checked`);
    }
}

export default check;
