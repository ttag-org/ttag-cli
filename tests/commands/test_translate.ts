import * as path from "path";
import { read, untranslatedStream } from "../../src/commands/translate";
import { serialize } from "../../src/lib/serializer";

const poPath = path.resolve(
    __dirname,
    "../fixtures/translateTest/translate.po"
);

test("translate poFile", () => {
    const poData = read(poPath);
    const stream = untranslatedStream(poData.translations);
    // skip first message(empty msgid in header)
    stream.next();
    stream.next("");
    stream.next(["test"]);
    stream.next(["test1", "test2"]);
    stream.next(["overwritten test1", "overwritten test2"]);
    const data = serialize(poData).toString();
    expect(data).toMatchSnapshot();
});
