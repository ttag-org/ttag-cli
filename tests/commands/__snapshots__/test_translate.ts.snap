// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translate poFile 1`] = `
"msgid \\"\\"
msgstr \\"\\"
\\"Content-Type: text/plain; charset=utf-8\\\\n\\"
\\"Plural-Forms: nplurals=2; plural=(n!=1);\\\\n\\"

#: tests/fixtures/checkTest/check-trans-exist.js:2
#: tests/fixtures/checkTest/check-trans-exist.js:3
msgid \\"test\\"
msgstr \\"test\\"

msgid \\"test_plural\\"
msgid_plural \\"test_plural\\"
msgstr[0] \\"test1\\"
msgstr[1] \\"test2\\"

msgid \\"filled_test\\"
msgstr \\"filled value\\"

msgid \\"partially_filled_test_plural\\"
msgid_plural \\"test_plural\\"
msgstr[0] \\"overwritten test1\\"
msgstr[1] \\"overwritten test2\\"

msgid \\"fully_filled_test_plural\\"
msgid_plural \\"test_plural\\"
msgstr[0] \\"filled value\\"
msgstr[1] \\"filled values\\"
"
`;
